### ISSUE TITLE
Web Sockets implementation test round 1

### ISSUE THREAD

**Issue by anithegregorian** [LACMTA/metro-api-v2#244]
Hi @albertkun @matikin9 ,

I tested the WebSocket implementation today, looks like it's still in the early stages and returning empty responses:

```
{"metadata": {"title": "Vehicle Positions"}, "type": "FeatureCollection", "features": []}
```

Perhaps this may clarify what could be an ideal implementation in FastAPI:

- socket should only emit events when vehicle position changes (to reduce operational load/cost on the client)
- moving marker plugin can be initialized with static/live data (preferably live data, but needs a fallback to static), which creates a designated route/path for the marker animation (which is the current approach), it only calls `pause()` and `resume()` methods randomly to simulate animation, it also limits changing the route/path
- for a live moving vehicle this would mean refactoring the moving marker code to progressively position the marker based on the current longitude/latitude emitted from the socket server (which is the correct approach)
- we should add longitude/latitude progressively for moving marker only after the socket is connected for every subscriber

You should check out the documentation for moving the marker as it says:

*Getter Methods*

 - ``` isRunning()```: return ```true``` if the marker is currently moving.
 - ```isPaused()```: return ```true``` if the marker is paused
 
 - ``` isEnded()```: return ```true``` if the marker is arrived to the last position or it has been stopped manually
 
 - ```isStarted()```: return ```true``` if the marker has started

**Note**: ```Marker.getLatLng()``` still works and give the current position
 
*Setter Methods*

 - ```start()```:  the marker begins its path or resumes if it is paused.
 - ``` stop()```: manually stops the marker, if you call ```start`` after, the marker starts again the polyline at the beginning.
 - ```pause()```: just pauses the marker
 - ``` resume()```: the marker resumes its animation
 - ```addLatLng(latlng, duration)```: adds a point to the polyline. Useful, if we have to set the path one by one.
 - ``` moveTo(latlng, duration)```: stops current animation and make the marker move to ```latlng``` in ```duration``` ms. 
 - ```addStation(pointIndex, duration)```: the marker will stop at the ```pointIndex```th points of the polyline during ```duration``` ms. You can't add a station at the first or last point of the polyline.

I'd be interested in `addLatLng(latlng, duration)` and maybe `addStation(pointIndex, duration)` if you have an idea of avg. time a vehicle would have a pitstop.

**Comment by matikin9** [comment]
What's the endpoint you're using?  We just pushed some updates yesterday and I'm getting data: 

`wss://api.metro.net/LACMTA/live/trip_detail/route_code/720?geojson=True`

![image](https://user-images.githubusercontent.com/1873072/228923964-4e7c123f-e727-4e97-89f2-f69ce8c4c154.png)

I'll need some time to go over the rest of your post.  I have no idea what it looks like to integrate the websocket connection on the frontend, so this is useful! :D

**Comment by anithegregorian** [comment]
@matikin9 its working now, wasn't showing up yesterday...

**Comment by anithegregorian** [comment]
@albertkun @matikin9 

I've run into one small issue with grouping bus routes in series. The API response returns an unsorted array of route codes, which makes it difficult to group. One of the route codes **SOFI** is also concerning because it's a string instead of an integer.

I was able to piece together a rather complex math expression to check for series:
```js
matchSeries(index){
	const current = store.agencies_bus_filtered[index];
	if (index === 0){
		return true
	}
	const previous = store.agencies_bus_filtered[index - 1];
	return Math.floor(parseInt(current.route_code) / 100) * 100 !== Math.floor(parseInt(previous.route_code) / 100) * 100
}
```

This goes at the template level with `v-if` to print the series label. Would it be possible to sort the response in ascending order by `route_code`?

Don't want to create new issues, as it's a small fix...

**Comment by albertkun** [comment]
@anithegregorian i have it currently running on `dev` here: 
https://dev-metro-api-v2.ofhq3vd1r7une.us-west-2.cs.amazonlightsail.com/

can you test the websocket there to see if the endpoint is ordered according to your request?

wss://dev-metro-api-v2.ofhq3vd1r7une.us-west-2.cs.amazonlightsail.com/LACMTA/live/trip_detail/route_code/720?geojson=True

Also, for this issue and going forward, please provide the exact endpoint you are testing with so we can troubleshoot correctly, thanks!

**Comment by anithegregorian** [comment]
@albertkun that would be the route overview (https://api.metro.net/LACMTA/route_overview?route_code=all) endpoint, haven't been able to test live API's yet

**Comment by albertkun** [comment]
Thank you for providing the example endpoint @anithegregorian !

I made a PR into `dev` can you check this in a couple minutes? I won't be able to respond until Wednesday though!

https://dev-metro-api-v2.ofhq3vd1r7une.us-west-2.cs.amazonlightsail.com/LACMTA/route_overview?route_code=all

**Comment by anithegregorian** [comment]
@albertkun I checked, thanks for the quick response.

When grouped the series looks like this from both `dev` and `prod` API:

![image](https://user-images.githubusercontent.com/143655/229792148-21c9c6a3-807b-43cf-a814-af1f44faef35.png)

Maybe this will give proper context. It would also make sense to group the route overview as a series in 100s, 200s, 300s, and so on, from FastAPI instead of returning a flat response.

Partly because I have to pair them up in the Petite Vue template and you cannot write JavaScripts within templates to jailbreak from loops. Because of Vue's data-driven approach, the API response may need refactoring (grouping by `route_code`):

```json
{
	"100": ["Route 14", "Route 25", "Route 35", "Route 99"],
	"200": ["Route 104", "Route 134", "Route 199"],
	"300": ["Route 301", "Route 333", "Route 378"]
}
```

Hope this helps...

**Comment by anithegregorian** [comment]
That's looking good @albertkun , the real challenge would be grouping series. I guess you're off today, will wait for further updates and keep moving with more template cleanup and schedule details page.

**Comment by albertkun** [comment]
@anithegregorian Thank you for the reply! Yeah, it was not ready for testing yet! It should be up on `dev` now:

https://dev-metro-api-v2.ofhq3vd1r7une.us-west-2.cs.amazonlightsail.com/all/route_overview

Also, regarding your comment about `route_code` by a string, we created a new field, `route_code_padded` which is an integer! Feel free to use that for sorting purposes!

Let me know if you have any questions!

**Comment by anithegregorian** [comment]
Hi @albertkun 

I tried with the refactored/additional field `route_code_padded` version API endpoint and it looks like this:

![image](https://user-images.githubusercontent.com/143655/232015521-de9c948c-aa8a-4087-b076-271dc46de2a6.png)

As you can see, grouping series won't work if the response is not sorted by `route_code_padded`, and would make sense to group them in series with the API response (instead of flat response)

```json
{
  "LACMTA": [
    {
      "100": [
        {
          "route_code": 2,
          "route_desc": "Westwood"
        },
        {
          "route_code": 56,
          "route_desc": "ASTRAL CITY"
        },
        {
          "route_code": 99,
          "route_desc": "BE SOFT."
        }
      ],
      "200": [
        {
          "route_code": 201,
          "route_desc": "Aboa"
        },
        {
          "route_code": 252,
          "route_desc": "CAPIO"
        },
        {
          "route_code": 298,
          "route_desc": "BUXUM"
        }
      ],
      "300": [
        {
          "route_code": 311,
          "route_desc": "Lutetia"
        },
        {
          "route_code": 344,
          "route_desc": "AONIDES"
        },
        {
          "route_code": 389,
          "route_desc": "FISCINA"
        }
      ]
    }
  ]
}
```

This is just an example of a response, that is grouped by series of `route_code`, and it's also sorted in each series. We can have a quick discussion if this doesn't make sense.

Thanks

**Comment by anithegregorian** [comment]
@albertkun is the WebSocket implementation direction aware? Meaning if LACMTA_Rail route 801 is moving in the opposite direction, is there any way to differentiate and update the map?

**Comment by albertkun** [comment]
@anithegregorian because the direction varies by route, you should probably use the `destination_code` to identify the direction of travel. We do not currently include `direction_id`.

**Comment by albertkun** [comment]
> Hi @albertkun 
> 
> I tried with the refactored/additional field `route_code_padded` version API endpoint and it looks like this:
> 
> ![image](https://user-images.githubusercontent.com/143655/232015521-de9c948c-aa8a-4087-b076-271dc46de2a6.png)
> 
> As you can see, grouping series won't work if the response is not sorted by `route_code_padded`, and would make sense to group them in series with the API response (instead of flat response)
> 
> ```json
> {
>   "LACMTA": [
>     {
>       "100": [
>         {
>           "route_code": 2,
>           "route_desc": "Westwood"
>         },
>         {
>           "route_code": 56,
>           "route_desc": "ASTRAL CITY"
>         },
>         {
>           "route_code": 99,
>           "route_desc": "BE SOFT."
>         }
>       ],
>       "200": [
>         {
>           "route_code": 201,
>           "route_desc": "Aboa"
>         },
>         {
>           "route_code": 252,
>           "route_desc": "CAPIO"
>         },
>         {
>           "route_code": 298,
>           "route_desc": "BUXUM"
>         }
>       ],
>       "300": [
>         {
>           "route_code": 311,
>           "route_desc": "Lutetia"
>         },
>         {
>           "route_code": 344,
>           "route_desc": "AONIDES"
>         },
>         {
>           "route_code": 389,
>           "route_desc": "FISCINA"
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> This is just an example of a response, that is grouped by series of `route_code`, and it's also sorted in each series. We can have a quick discussion if this doesn't make sense.
> 
> Thanks

Sure, I think a quick discussion would be helpful! What is your availability on Monday? I know it is getting late in India, haha 😅

**Comment by anithegregorian** [comment]
I can do it between 10.00 - 10.30 AM PDT on Monday

**Comment by albertkun** [comment]
> I can do it between 10.00 - 10.30 AM PDT on Monday

Perfect, lets do that! Thank you!

**Comment by anithegregorian** [comment]
@albertkun I'd also like to go over the frequency of WebSocket broadcasts, I did a barebones test, opened up a WebSocket connection to the 801 route, and waited for Firefox to collect resource information.

If I leave the browser window open for 20-25 mins the interaction time in Browser gets sketchy (there's a slight delay in response), and we'd have to go over this as well. Just giving you a head up.

It's also been highlighted in my first response on this thread:

- Socket should only emit events when vehicle position changes (to reduce operational load/cost on the client)

Have a great weekend!

**Comment by anithegregorian** [comment]
Hey @albertkun 

It seems like the live/dev API servers are down, live API was down since yesterday, getting this error:

```
503 Service Temporarily Unavailable
```

**Comment by albertkun** [comment]
@anithegregorian yeah, not sure how or why the api went down all of a sudden, but looking into it. i'm pushing a new version to `dev` right now #267 , with reverting some previous changes.

**Comment by albertkun** [comment]
@anithegregorian api should be back up now! Will monitor throughout the day for any issues!

**Comment by anithegregorian** [comment]
Thanks @albertkun , think you'll need some sort of robust monitoring if this thing goes down once it goes live

**Comment by albertkun** [comment]
@anithegregorian I made a quick change to the time out on `dev`! Let me know if it's working again and I'll merge to `dev`!

**Comment by albertkun** [comment]
@anithegregorian I identified the issue, the socket was/closing because you were most likely using `geojson=True` as the parameter. Please set that to `false`. If you need the `geojson` format, I'll go ahead and fix it.

**Comment by anithegregorian** [comment]
@albertkun Yes please, the entire codebase is parsing through GeoJSON since it's standard/portable across any mapping platform, will keep the broken mapping link in mind.

**Comment by albertkun** [comment]
@anithegregorian Ok, sounds good! i'll get around to adding that functionality on Friday (Saturday your time)! Sorry for the delay!

**Comment by albertkun** [comment]
@anithegregorian Should be up and fixed now! Let me know if you have any questions!

**Comment by anithegregorian** [comment]
@albertkun As we close in on this, two things that I noticed:

- Server sockets are still closing connections
- Is it possible to send a `LineString` in GeoJSON response for routes (currently its built with lng/ltd which looks ugly)

Rest looks steadfast! In conclusion, it would also benefit to watch out for any downtime in API service and how to mitigate them on the front end.

**Comment by albertkun** [comment]
Thank you @anithegregorian ! 

- Unfortunately, yes, the server sockets are still closing.. I'm not sure why- except that it could still be related to a timeout setting...
- For the `LineString` which GeoJSON endpoint would you like it for? The `routes`. I'm not sure if I can switch the formatting for the GeoJSON, but will try!

Rest does seem pretty good I think. Definitely, I'm still having trouble working with websockets (as you can see!!).
